File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.java
// Last updated: 2/9/2026, 9:54:03 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int totalCount = 0;
    public int goodNodes(TreeNode root) {
        if(root == null) return 0;
        findCount(root, root.val);
        return totalCount;
    }

    private void findCount(TreeNode root, int max) {
        if(root == null) return;
        if(root.val >= max) {
            totalCount++;
            findCount(root.right, root.val);
            findCount(root.left, root.val);
        }
        else {
            findCount(root.right, max);
            findCount(root.left, max);
        }
    }
}

File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.py
# Last updated: 2/9/2026, 9:54:00 PM
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        output = 0
        maxVal = float('-inf')
        if not root:
            return output
        
        def findGoodNodes(root, maxVal):
            output = 0

            if not root:
                return output

            maxVal = max(maxVal, root.val)
            if (maxVal == root.val):
                output += 1

            r = findGoodNodes(root.right, maxVal) 
            l = findGoodNodes(root.left, maxVal)

            return r + l + output
        
        return findGoodNodes(root, maxVal)
            

        